### PR TITLE
chore: refactor StaticPod and StaticPodStatus into typed.Resource

### PR DESCRIPTION
**Pull request by DmitriyMV**
This two required some additional attention and were split into separate branch from #5472.

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

---
This change is [Reviewable](https://reviewable.io/reviews/siderolabs/talos/5497)

### CODE REVIEW

**Comment by DmitriyMV** [comment]
So the root problem was that old `StaticPodSpec.MarshalYAML` was defined on pointer receiver. That is
```
func (spec *StaticPodSpec) MarshalYAML() (interface{}, error) {
```

With typed.Resource the new code `Spec() interface{}` returns `StaticPodSpec{}` and not `*StaticPodSpec{}` which is bad if you are trying to call any methods on returned `interface{}` using type assertion, since you can't call methods which expect pointer receiver on non-pointer interface.

Minimal repro/example: https://go.dev/play/p/PLxBYfXoLCf

**Comment by frezbo** [comment]
good catch

**Comment by DmitriyMV** [comment]
> good catch

Big thanks to @smira for saying that `MarshalYAML` happen on empty interface.

File: pkg/machinery/resources/network/node_address.go [hunk 1]
```diff
@@ -42,7 +42,9 @@ type NodeAddressSpec struct {
 
 // DeepCopy generates a deep copy of NodeAddressSpec.
 func (spec NodeAddressSpec) DeepCopy() NodeAddressSpec {
-	return spec
+	return NodeAddressSpec{
+		Addresses: append([]netaddr.IPPrefix(nil), spec.Addresses...),
```
Comment (smira) [inline pkg/machinery/resources/network/node_address.go:46]: :+1:

**Review by smira — APPROVED** [review]

**Review by frezbo — APPROVED** [review]

**Pull request merged**